core/flyout.js

```diff
@@ -259,47 +315,144 @@ Blockly.Flyout.prototype.position = function() {
 
-  // Record the height for Blockly.Flyout.getMetrics_.
-  this.height_ = metrics.viewHeight;
+  // Record the height for Blockly.Flyout.getMetrics_, or width if the layout is
+  // horizontal.
+  if (this.horizontalLayout_) {
+    this.width_ = metrics.viewWidth;
+  } else {
+    this.height_ = metrics.viewHeight;
+  }
 
   // Update the scrollbar (if one exists).
   if (this.scrollbar_) {
     this.scrollbar_.resize();
   }
+  // The blocks need to be visible in order to be laid out and measured
+  // correctly, but we don't want the flyout to show up until it's properly
+  // sized.  Opacity is set to zero in show().
+  this.svgGroup_.style.opacity = 1;
+};
+
+/**
+ * Create and set the path for the visible boundaries of the toolbox.
+ * @param {number} width The width of the toolbox, not including the
+ *     rounded corners.
+ * @param {number} height The height of the toolbox, not including
+ *     rounded corners.
+ * @private
+ */
+Blockly.Flyout.prototype.setBackgroundPath_ = function(width, height) {
+  if (this.horizontalLayout_) {
+    this.setBackgroundPathHorizontal_(width, height);
+  } else {
+    this.setBackgroundPathVertical_(width, height);
+  }
+};
+
+/**
+ * Create and set the path for the visible boundaries of the toolbox in vertical mode.
+ * @param {number} width The width of the toolbox, not including the
+ *     rounded corners.
+ * @param {number} height The height of the toolbox, not including
+ *     rounded corners.
+ * @private
+ */
+Blockly.Flyout.prototype.setBackgroundPathVertical_ = function(width, height) {
+    var atRight = this.toolboxPosition_ == Blockly.TOOLBOX_AT_RIGHT;
+  // Decide whether to start on the left or right.
+  var path = ['M ' + (atRight ? this.width_ : 0) + ',0'];
+  // Top.
+  path.push('h', width);
+  // Rounded corner.
+  path.push('a', this.CORNER_RADIUS, this.CORNER_RADIUS, 0, 0,
+      atRight ? 0 : 1,
+      atRight ? -this.CORNER_RADIUS : this.CORNER_RADIUS,
+      this.CORNER_RADIUS);
+  // Side closest to workspace.
+  path.push('v', Math.max(0, height - this.CORNER_RADIUS * 2));
+  // Rounded corner.
+  path.push('a', this.CORNER_RADIUS, this.CORNER_RADIUS, 0, 0,
+      atRight ? 0 : 1,
+      atRight ? this.CORNER_RADIUS : -this.CORNER_RADIUS,
+      this.CORNER_RADIUS);
+  // Bottom.
+  path.push('h', -width);
+  path.push('z');
+  this.svgBackground_.setAttribute('d', path.join(' '));
+};
+
+/**
+ * Create and set the path for the visible boundaries of the toolbox in horizontal mode.
+ * @param {number} width The width of the toolbox, not including the
+ *     rounded corners.
+ * @param {number} height The height of the toolbox, not including
+ *     rounded corners.
+ * @private
+ */
+Blockly.Flyout.prototype.setBackgroundPathHorizontal_ = function(width, height) {
+    var atTop = this.toolboxPosition_ == Blockly.TOOLBOX_AT_TOP;
+  // start at top left.
+  var path = ['M 0,' + (atTop ? 0 : this.CORNER_RADIUS)];
+
+  if (atTop) {
+    // top
```

Comment:
Top.  To make consistent with properly capitalized and punctuated comments in previous function.  Fix throughout this function.